pennylane/qnode.py

```diff
@@ -184,10 +184,22 @@ def __init__(
         self.qfunc_output = None
         # store the user-specified differentiation method
         self.diff_method = diff_method
+        self.diff_method_change = False
 
         self._tape, self.interface, self.device, tape_diff_options = self.get_tape(
             device, interface, diff_method
         )
+        # Update diff_method to reflect which method has been selected
+        if self.diff_method == "best":
+            self.diff_method_change = True
+            if tape_diff_options["method"] == "device":
+                self.diff_method = "device"
+            elif tape_diff_options["method"] == "backprop":
+                self.diff_method = "backprop"
+            elif tape_diff_options["method"] == "best":
+                self.diff_method = "parameter-shift"

```

Comment:
Here we could actually fall back to finite-differences for certain gates (as per [this comment](https://github.com/PennyLaneAI/pennylane/pull/1560#issuecomment-902739358)).

See the example of a custom gate that defines `grad_method = "F"`:
```python
import pennylane as qml
import numpy as np

class MyRX(qml.operation.Operation):
    num_params = 1
    num_wires = 1
    par_domain = "R"
    is_composable_rotation = True
    basis = "X"
    grad_method = "F"

    @classmethod
    def _matrix(cls, *params):
        return qml.RX._matrix(*params)

dev = qml.device('lightning.qubit', wires=3)
dev.operations.add("MyRX")

@qml.qnode(dev, diff_method='best')
def circuit(x):
    MyRX(x, wires=1)
    return qml.expval(qml.PauliZ(1))

par = qml.numpy.array(0.3)
qml.grad(circuit)(par)
```
Here we have `diff_method='best'`. Yet, as `MyRX` only supports finite differences and `tape_diff_options["method"] == "best"`, as per the current logic we'd set `self.diff_method = "parameter-shift"`. This could be confusing for users, as they might think that we've used the parameter-shift rule to compute the gradient of the circuit wrt. to the parameter of `MyRX`.

----

This case might have to be determined later in the pipeline, once `self.qtape` has been created.

We might need some logic like

```python
finite_diff = any([x.get('grad_method', None) == "F" for x in self.qtape._par_info.values()])
```
to tell if we've fallen back to finite differences somewhere. If that's the case, then maybe setting `self.diff_method = "finite-diff"` would be the way, to show users that we've fallen back to finite differences at least for one gate.

Reply:
I can not run this code. I am getting:
```
Traceback (most recent call last):
  .
  .
    qml.grad(circuit)(par)
  File "\pennylane\_grad.py", line 101, in __call__
    grad_value, ans = self._get_grad_fn(args)(*args, **kwargs)
  File "\autograd\wrap_util.py", line 20, in nary_f
    return unary_operator(unary_f, x, *nary_op_args, **nary_op_kwargs)
  File "\pennylane\_grad.py", line 118, in _grad_with_forward
    vjp, ans = _make_vjp(fun, x)
  File "\autograd\core.py", line 10, in make_vjp
    end_value, end_node =  trace(start_node, fun, x)
  File "\autograd\tracer.py", line 10, in trace
    end_box = fun(start_box)
  File "\autograd\wrap_util.py", line 15, in unary_f
    return fun(*subargs, **kwargs)
  File "\pennylane\qnode.py", line 602, in __call__
    res = self.qtape.execute(device=self.device)
  File "\pennylane\tape\tape.py", line 1277, in execute
    return self._execute(params, device=device)
  File "\autograd\tracer.py", line 44, in f_wrapped
    ans = f_wrapped(*argvals, **kwargs)
  File "\autograd\tracer.py", line 48, in f_wrapped
    return f_raw(*args, **kwargs)
  File "\pennylane\interfaces\autograd.py", line 165, in _execute
    res = self.execute_device(params, device=device)
  File "\pennylane\tape\tape.py", line 1308, in execute_device
    res = device.execute(self)
  File "\pennylane\_qubit_device.py", line 194, in execute
    self.apply(circuit.operations, rotations=circuit.diagonalizing_gates, **kwargs)
  File "\pennylane_lightning\lightning_qubit.py", line 112, in apply
    self._pre_rotated_state = self.apply_lightning(self._state, operations)
  File "\pennylane_lightning\lightning_qubit.py", line 140, in apply_lightning
    apply(state_vector, op_names, op_wires, op_inverse, op_param)
IndexError: invalid unordered_map<K, T> key
```

Reply:
Hi @ankit27kh, could you try upgrading PennyLane-Lightning via `pip install --upgrade pennylane-lightning`?

If that doesn't come around could you:
1. Open an [issue in the PennyLane-Lightning GitHub repository](https://github.com/PennyLaneAI/pennylane-lightning/issues)
2. If you have the released version of PennyLane-Cirq, try the following:
```python
import pennylane as qml
import numpy as np

class MyRX(qml.operation.Operation):
    num_params = 1
    num_wires = 1
    par_domain = "R"
    is_composable_rotation = True
    basis = "X"
    grad_method = "F"

    @classmethod
    def _matrix(cls, *params):
        return qml.RX._matrix(*params)

dev = qml.device('cirq.simulator', wires=3, shots=None)
dev._operation_map["MyRX"] = dev._operation_map["RX"]

@qml.qnode(dev, diff_method='best')
def circuit(x):
    MyRX(x, wires=1)
    return qml.expval(qml.PauliZ(1))

par = qml.numpy.array(0.3)
qml.grad(circuit)(par)
```

Reply:
Thanks @ankit27kh! Try adding the `MyRX` gate before setting it (`dev.operations.add("MyRX")`) when attempting to run the Cirq plugin example above. It believe it should work then.

Reply:
Still the same error.

I am getting the specs and can draw the circuit too.
```
circuit.specs
>>>{'gate_sizes': defaultdict(int, {1: 1}),
 'gate_types': defaultdict(int, {'MyRX': 1}),
 'num_operations': 1,
 'num_observables': 1,
 'num_diagonalizing_gates': 0,
 'num_used_wires': 1,
 'depth': 1,
 'num_trainable_params': 1,
 'num_parameter_shift_executions': 1,
 'num_device_wires': 3,
 'device_name': 'cirq.simulator',
 'diff_method': 'best'}
```
But if I try to run it, it fails.

Reply:
Hi @ankit27kh, my bad! Managed to reproduce your error with the Cirq plugin, and this change now helps with executing the example. Could you try this (adding support for the new operation with `dev._complete_operation_map["MyRX"] = dev._operation_map["RX"]`):
```python
import pennylane as qml
import numpy as np

class MyRX(qml.operation.Operation):
    num_params = 1
    num_wires = 1
    par_domain = "R"
    is_composable_rotation = True
    basis = "X"
    grad_method = "F"

    @classmethod
    def _matrix(cls, *params):
        return qml.RX._matrix(*params)

dev = qml.device('cirq.simulator', wires=3, shots=None)
dev._complete_operation_map["MyRX"] = dev._complete_operation_map["RX"]

@qml.qnode(dev, diff_method='best')
def circuit(x):
    MyRX(x, wires=1)
    return qml.expval(qml.PauliZ(1))

par = qml.numpy.array(0.3)
qml.grad(circuit)(par)
```

Reply:
@antalszava it still doesn't work.

if this works for you can me tell me the output of this:
```
import pennylane as qml
import numpy as np

class MyRX(qml.operation.Operation):
    num_params = 1
    num_wires = 1
    par_domain = "R"
    is_composable_rotation = True
    basis = "X"
    grad_method = "F"

    @classmethod
    def _matrix(cls, *params):
        return qml.RX._matrix(*params)

dev = qml.device('cirq.simulator', wires=3, shots=None)
dev._complete_operation_map["MyRX"] = dev._complete_operation_map["RX"]

@qml.qnode(dev, diff_method='best')
def circuit(x):
    MyRX(x, wires=1)
    return qml.expval(qml.PauliZ(1))

print(circuit)
par = qml.numpy.array(0.3)
qml.grad(circuit)(par)
circuit(2.2)
print(circuit)
```

Reply:
Hi @ankit27kh, thanks for bearing with me on this! Let's switch to the core of PennyLane such that there are no reproducibility issues.

Could you check the following with PennyLane v0.17.0? It changes the `grad_method` of `qml.RX` to `'F'` (finite-differences) and compares getting the gradient of the circuit with `diff_method="best"` and `diff_method="parameter-shift"`.
```python
import pennylane as qml
import numpy as np

dev = qml.device('default.qubit', wires=3, shots=None)

qml.RX.grad_method = 'F'
print("The new grad method for RX is: ", qml.RX.grad_method)

def circuit(x):
    qml.RX(x, wires=1)
    return qml.expval(qml.PauliZ(1))


qnode1 = qml.QNode(circuit, dev, diff_method='best')

par = qml.numpy.array(0.3)
qml.grad(qnode1)(par)

# Sanity check
qnode2 = qml.QNode(circuit, dev, diff_method='parameter-shift')

try:
    qml.grad(qnode2)(par)
except ValueError:
    print("RX doesn't support the parameter-shift rule")
```
The expected output would be:
```
The new grad method for RX is:  F
RX doesn't support the parameter-shift rule
```

Reply:
Yes, I am getting this output.